File: go/delete-nodes-and-return-forest.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 func CreateForest(node *TreeNode, toDelete map[int]bool, forest *[]*TreeNode) bool{
    if node == nil {
        return false
    }
    if CreateForest(node.Left, toDelete, forest) {
        node.Left = nil
    }
    if CreateForest(node.Right, toDelete, forest) {
        node.Right = nil
    }
    if toDelete[node.Val] == true {
        if node.Left != nil {
            *(forest) = append(*(forest), node.Left)
        }
        if node.Right != nil {
            *(forest) = append(*(forest), node.Right)
        }
        return true
    }
    return false
}

func delNodes(root *TreeNode, to_delete []int) []*TreeNode {
    forest := make([]*TreeNode, 0)
    toDelete := make(map[int]bool)
    for i := 0; i < len(to_delete); i++ {
        toDelete[to_delete[i]] = true
    }
    deleteRoot := CreateForest(root, toDelete, &forest)
    if deleteRoot == false {
        forest = append(forest, root)
    }
    //fmt.Println(forest)
    return forest
}

File: python/delete-nodes-and-return-forest.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def create_forest(node: Optional[TreeNode], to_delete_set: Set[int], forest: List[TreeNode]) -> bool:
        if node == None:
            return False
        if Solution.create_forest(node.left, to_delete_set, forest):
            node.left = None
        if Solution.create_forest(node.right, to_delete_set, forest):
            node.right = None
        if node.val in to_delete_set:
            if node.left != None:
                forest.append(node.left)
            if node.right != None:
                forest.append(node.right)
            return True
        return False

    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        to_delete_set = {item for item in to_delete}
        forest = []
        delete_root = Solution.create_forest(root, to_delete_set, forest)
        if delete_root == False:
            forest.append(root)
        return forest
